kubeadm/scripts/PrepareNode.ps1

```diff
@@ -59,81 +59,99 @@ $global:PowershellArgs = "-ExecutionPolicy Bypass -NoProfile"
-    Import-Module "c:\k\hns.psm1"
-    # TODO(marosset): check if network already exists before creatation
-    New-HnsNetwork -Type NAT -Name nat
+if (-not (Test-Path "$global:KubernetesPath\hns.psm1")) {
+    DownloadFile "$global:KubernetesPath\hns.psm1" https://github.com/Microsoft/SDN/raw/master/Kubernetes/windows/hns.psm1
+} else {
+    Write-Host "hns.psm1 has been already downloaded"
 }
 
-Write-Host "Registering wins service"
-wins.exe srv app run --register
-start-service rancher-wins
-
-mkdir -force C:\var\log\kubelet
-mkdir -force C:\var\lib\kubelet\etc\kubernetes
-mkdir -force C:\etc\kubernetes\pki
-New-Item -path C:\var\lib\kubelet\etc\kubernetes\pki -type SymbolicLink -value C:\etc\kubernetes\pki\
-
-$StartKubeletFileContent = '$FileContent = Get-Content -Path "/var/lib/kubelet/kubeadm-flags.env"
-$global:KubeletArgs = $FileContent.TrimStart(''KUBELET_KUBEADM_ARGS='').Trim(''"'')
-
-$global:containerRuntime = {{CONTAINER_RUNTIME}}
-
-if ($global:containerRuntime -eq "Docker") {
-    $netId = docker network ls -f name=host --format "{{ .ID }}"
-
-    if ($netId.Length -lt 1) {
-    docker network create -d nat host
-    }
+if (-not (Get-Service rancher-wins -ErrorAction Ignore)) {
+    DownloadFile "$global:KubernetesPath\wins.exe" https://github.com/rancher/wins/releases/download/v0.0.4/wins.exe
+    Write-Host "Registering rancher-wins service"
+    wins.exe srv app run --register
+    Start-Service rancher-wins
+} else {
+    Write-Host "rancher-wins has been already installed"
 }
 
-$cmd = "C:\k\kubelet.exe $global:KubeletArgs --cert-dir=$env:SYSTEMDRIVE\var\lib\kubelet\pki --config=/var/lib/kubelet/config.yaml --bootstrap-kubeconfig=/etc/kubernetes/bootstrap-kubelet.conf --kubeconfig=/etc/kubernetes/kubelet.conf --hostname-override=$(hostname) --pod-infra-container-image=`"mcr.microsoft.com/oss/kubernetes/pause:1.4.1`" --enable-debugging-handlers --cgroups-per-qos=false --enforce-node-allocatable=`"`" --network-plugin=cni --resolv-conf=`"`" --log-dir=/var/log/kubelet --logtostderr=false --image-pull-progress-deadline=20m"
+if (-not (Test-Path "$global:NssmInstallDirectory\nssm.exe")) {
+    Write-Host "Installing nssm"
+    $arch = "win32"
+    if ([Environment]::Is64BitOperatingSystem) {
+        $arch = "win64"
+    }
 
-Invoke-Expression $cmd'
-$StartKubeletFileContent = $StartKubeletFileContent -replace "{{CONTAINER_RUNTIME}}", "`"$ContainerRuntime`""
-Set-Content -Path $global:StartKubeletScript -Value $StartKubeletFileContent
+    mkdir -Force $global:NssmInstallDirectory | Out-Null
+    DownloadFile nssm.zip https://k8stestinfrabinaries.blob.core.windows.net/nssm-mirror/nssm-2.24.zip
+    tar C $global:NssmInstallDirectory -xvf .\nssm.zip --strip-components 2 */$arch/*.exe
+    Remove-Item -Force .\nssm.zip
+
+    $env:path += ";$global:NssmInstallDirectory"
+    $newPath = "$global:NssmInstallDirectory;" +
+    [Environment]::GetEnvironmentVariable("PATH", [EnvironmentVariableTarget]::Machine)
```

Comment:
This will prefix system path with the nssm install directory. For now this shouldn't be an issue. But it is bad practice, as it will slow down dynamic dll imports as this directory will get checked first every time. Also it will allow system executables and libraries to be accidentally overwritten.

```suggestion
    $newPath = [Environment]::GetEnvironmentVariable("PATH", [EnvironmentVariableTarget]::Machine) + ";$global:NssmInstallDirectory"
```

Reply:
It makes sense.